### PR TITLE
Cyphers/pattern

### PR DESCRIPTION
**Pull request by diyessi**
* Match based on outputs rather than nodes
* Add `Or` and `Branch`
* Initial support for recurrent pattern (as opposed to separate matcher)
* Almost completely compatible with existing patterns

### CODE REVIEW

File: src/ngraph/CMakeLists.txt [hunk 1]
```diff
@@ -559,11 +559,22 @@ set (SRC
     pass/pass_util.cpp
     pattern/matcher.cpp
     pattern/matcher.hpp
+    pattern/op/any.cpp
     pattern/op/any.hpp
+    pattern/op/any_of.cpp
     pattern/op/any_of.hpp
+    pattern/op/branch.cpp
+    pattern/op/branch.hpp
```
Comment (ayzhuang) [inline src/ngraph/CMakeLists.txt]: Add pattern/op/capture.hpp and pattern/op/capture.cpp here.

File: src/ngraph/pattern/op/capture.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                {
+                    m_static_nodes = static_nodes;
+                }
+
+                virtual bool match_value(pattern::Matcher* matcher,
+                                         const Output<Node>& pattern_value,
+                                         const Output<Node>& graph_value) override;
+
+            protected:
+                std::set<Node*> m_static_matches;
```
Comment (ayzhuang) [inline src/ngraph/pattern/op/capture.hpp]: should be m_static_nodes.

File: src/ngraph/pattern/op/capture.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+const NodeTypeInfo& pattern::op::Capture::get_type_info() const
+{
+    return type_info;
+}
+
+bool pattern::op::Capture::match_value(Matcher* matcher,
+                                       const Output<Node>& pattern_value,
+                                       const Output<Node>& graph_value)
+{
+    matcher->capture();
```
Comment (ayzhuang) [inline src/ngraph/pattern/op/capture.cpp]: matcher->capture(m_static_nodes)?

File: src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp [hunk 4]
```diff
@@ -540,8 +541,12 @@ void ngraph::runtime::cpu::pass::RNNFusion::construct_rnn_lstm_fprop()
...
     auto lstm_goe = std::make_shared<ngraph::op::GetOutputElement>(lstm, 1);
     // We cannot attach labels to multi-output nodes, so we attach a label to the goe instead
-    auto lstm_goe_label =
-        std::make_shared<pattern::op::Label>(lstm_goe, nullptr, NodeVector{lstm_goe});
+    auto lstm_goe_label = std::make_shared<pattern::op::Label>(
+        lstm_goe,
+        nullptr,
+        OutputVector{std::make_shared<pattern::op::Or>(
+            OutputVector{lstm_goe, std::make_shared<ngraph::op::GetOutputElement>(lstm, 0)})});
+    // NodeVector{lstm_goe, std::make_shared<ngraph::op::GetOutputElement>(lstm, 0)});
```
Comment (ayzhuang) [inline src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp]: remove comment.

File: src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp [hunk 5]
```diff
@@ -540,8 +541,12 @@ void ngraph::runtime::cpu::pass::RNNFusion::construct_rnn_lstm_fprop()
                                                    ref_rnn_type);
     auto lstm_goe = std::make_shared<ngraph::op::GetOutputElement>(lstm, 1);
     // We cannot attach labels to multi-output nodes, so we attach a label to the goe instead
-    auto lstm_goe_label =
-        std::make_shared<pattern::op::Label>(lstm_goe, nullptr, NodeVector{lstm_goe});
+    auto lstm_goe_label = std::make_shared<pattern::op::Label>(
+        lstm_goe,
+        nullptr,
+        OutputVector{std::make_shared<pattern::op::Or>(
```
Comment (ayzhuang) [inline src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp]: Why do we need Or here? Aren't we always slice lstm_goe?

Comment (diyessi) [reply src/ngraph/runtime/cpu/pass/cpu_rnn_fusion.cpp]: The last cell is connected to the graph by a different output than the output used for cell to cell connections.

**Pull request merged**